pkg/ride/environment.go

```diff
@@ -1136,25 +1043,31 @@ func NewEnvironmentWithWrappedState(
 			return nil, err
 		}
 		if senderBalance < payment.Amount {
-			return nil, errors.New("not enough money for tx attached payments")
-		}
-		searchBalance, searchAddr, err := st.diff.findBalance(recipient, payment.Asset)
-		if err != nil {
-			return nil, errors.Wrap(err, "failed to create RIDE environment with wrapped state")
-		}
-		err = st.diff.changeBalance(searchBalance, searchAddr, int64(payment.Amount), payment.Asset, recipient)
-		if err != nil {
-			return nil, errors.Wrap(err, "failed to create RIDE environment with wrapped state")
-		}
-
-		senderSearchBalance, senderSearchAddr, err := st.diff.findBalance(callerRcp, payment.Asset)
-		if err != nil {
-			return nil, errors.Wrap(err, "failed to create RIDE environment with wrapped state")
+			return nil, errors.Errorf("not enough money for tx attached payment #%d of asset '%s' with amount %d",
+				i+1, payment.Asset.String(), payment.Amount)
 		}
-
-		err = st.diff.changeBalance(senderSearchBalance, senderSearchAddr, -int64(payment.Amount), payment.Asset, callerRcp)
-		if err != nil {
-			return nil, errors.Wrap(err, "failed to create RIDE environment with wrapped state")
+		if payment.Asset.Present {
```

Comment:
It seems like this logic `if payment.Asset.Present ; else` repeats, I also suggest moving it into a different method/function

Reply:
It was like this, I don't like it. In case of joining this functions into one we have to use `OptionaAsset`, but I wan't to keep diff state simple and assets and waves separated.